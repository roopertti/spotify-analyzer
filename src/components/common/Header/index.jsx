import React from 'react';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCaretDown } from '@fortawesome/free-solid-svg-icons';

import LinkButton from '../LinkButton';

const Brand = styled.h1`
  display: inline-flex;
  margin: 0;
  font-family: 'Raleway', sans-serif;
  font-style: normal;
  font-weight: bold;
  font-size: 2rem;
  box-shadow: 0px 4px 30px 0px #b118c8 30%;
  background-color: #2f29bf;
  background-image: linear-gradient(90.05deg, #2f29bf -15.19%, rgba(255, 255, 255, 0) 46.55%),
    linear-gradient(90.13deg, #b118c8 46.55%, rgba(255, 255, 255, 0) 142.37%), linear-gradient(0deg, #ff8600, #ff8600);
  background-size: 100%;
  -webkit-background-clip: text;
  -moz-background-clip: text;
  -webkit-text-fill-color: transparent;
  -moz-text-fill-color: transparent;

  @media (max-width: 960px) {
    font-size: 1.8rem;
  }

  @media (max-width: 600px) {
    font-size: 1.5rem;
  }
`;

const StyledHeader = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 3rem;
  box-shadow: 0px 10px 30px rgba(47, 41, 191, 0.3), 0px -10px 30px rgba(177, 24, 200, 0.3);

  @media (max-width: 960px) {
    padding: 2rem;
  }

  @media (max-width: 600px) {
    padding: 1.5rem;
  }
`;

const Backdrop = styled.div`
  position: fixed;
  z-index: 2;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  backdrop-filter: blur(12px);
`;

const MobileNav = styled.nav`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  padding: 6rem 3rem 0 0;

  @media (max-width: 960px) {
    padding: 6rem 2rem 0 0;
  }

  @media (max-width: 600px) {
    padding: 6rem 1.5rem 0 0;
  }

  button {
    margin-bottom: 1.5rem;
  }

  a {
    margin-bottom: 1.5rem;
  }
`;

const MenuButton = styled.button`
  font-size: 2rem;
  outline: none;
  z-index: 3;
  background: none;
  border-radius: 50%;
  width: 2.5rem;
  height: 2.5rem;
  color: #b118c8;
  border: none;

  transform: ${(props) => (props.flipped ? 'rotate(180deg)' : 'none')};
  transition-duration: 0.3s;

  &:hover {
    background-color: #e5e5e5;
  }

  &:focus: {
    box-shadow: linear-gradient(90.2deg, #2f29bf 0%, rgba(255, 255, 255, 0) 100%), #b118c8;
  }

  @media (max-width: 960px) {
    font-size: 1.8rem;
  }

  @media (max-width: 600px) {
    font-size: 1.5rem;
  }
`;

export default function Header() {
  const [mobileMenuOpen, setMobileMenuOpen] = React.useState(false);

  return (
    <StyledHeader>
      <Brand>Spotify analyzer</Brand>
      <MenuButton data-testid="menu-button" flipped={mobileMenuOpen}>
        <FontAwesomeIcon icon={faCaretDown} onClick={() => setMobileMenuOpen(!mobileMenuOpen)} />
      </MenuButton>
      {mobileMenuOpen && (
        <Backdrop>
          <MobileNav data-testid="mobile-menu">
            <LinkButton className="link" href="#">
              Authorize
            </LinkButton>
            <LinkButton className="link" href="#">
              Features
            </LinkButton>
            <LinkButton className="link" href="#">
              Github
            </LinkButton>
          </MobileNav>
        </Backdrop>
      )}
    </StyledHeader>
  );
}
