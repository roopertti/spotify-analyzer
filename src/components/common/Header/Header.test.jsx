import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { screen, fireEvent, render, act } from '@testing-library/react';

import Header from './index';

describe('Header tests', () => {
  it('renders Header', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render Header, close by backdrop click', async () => {
    const { getByTestId } = render(<Header />);
    const menuButton = getByTestId('menu-button');
    fireEvent.click(menuButton);
    expect(screen.getByTestId('mobile-menu')).toBeTruthy();
    fireEvent.click(screen.getByText('Authorize'));
    expect(screen.getByTestId('mobile-menu')).toBeTruthy();
    await act(async () => {
      fireEvent.click(screen.getByTestId('backdrop'));
      await new Promise((r) => setTimeout(r, 1000));
    });
    expect(screen.queryByTestId('mobile-menu')).toBeFalsy();
  });

  it('render Header, close by menu button click', async () => {
    const { getByTestId } = render(<Header />);
    const menuButton = getByTestId('menu-button');
    fireEvent.click(menuButton);
    expect(screen.getByTestId('mobile-menu')).toBeTruthy();
    fireEvent.click(menuButton);
    await new Promise((r) => setTimeout(r, 1000));
    expect(screen.queryByTestId('mobile-menu')).toBeFalsy();
  });
});
