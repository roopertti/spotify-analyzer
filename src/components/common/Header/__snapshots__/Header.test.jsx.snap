// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests renders Header 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 0;
  font-family: 'Raleway',sans-serif;
  font-style: normal;
  font-weight: bold;
  font-size: 2rem;
  box-shadow: 0px 4px 30px 0px #b118c8 30%;
  background-color: #2f29bf;
  background-image: linear-gradient(90.05deg,#2f29bf -15.19%,rgba(255,255,255,0) 46.55%), linear-gradient(90.13deg,#b118c8 46.55%,rgba(255,255,255,0) 142.37%),linear-gradient(0deg,#ff8600,#ff8600);
  background-size: 100%;
  -webkit-background-clip: text;
  -moz-background-clip: text;
  -webkit-text-fill-color: transparent;
  -moz-text-fill-color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 3rem;
  box-shadow: 0px 10px 30px rgba(47,41,191,0.3),0px -10px 30px rgba(177,24,200,0.3);
}

.c2 {
  font-size: 2rem;
  outline: none;
  z-index: 3;
  background: none;
  border-radius: 50%;
  width: 2.5rem;
  height: 2.5rem;
  color: #b118c8;
  border: none;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
}

.c2:hover {
  background-color: #e5e5e5;
}

.c2:focus: {
  box-shadow: linear-gradient(90.2deg,#2f29bf 0%,rgba(255,255,255,0) 100%),#b118c8;
}

@media (max-width:960px) {
  .c1 {
    font-size: 1.8rem;
  }
}

@media (max-width:600px) {
  .c1 {
    font-size: 1.5rem;
  }
}

@media (max-width:960px) {
  .c0 {
    padding: 2rem;
  }
}

@media (max-width:600px) {
  .c0 {
    padding: 1.5rem;
  }
}

@media (max-width:960px) {
  .c2 {
    font-size: 1.8rem;
  }
}

@media (max-width:600px) {
  .c2 {
    font-size: 1.5rem;
  }
}

<header
  className="c0"
>
  <h1
    className="c1"
  >
    Spotify analyzer
  </h1>
  <button
    className="c2"
    data-testid="menu-button"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-caret-down fa-w-10 "
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      onClick={[Function]}
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</header>
`;
